File: configuration-manager-go-client/configuration.go
package configmanager

import (
	"encoding/json"
	"time"
)

// Configuration represents a configuration entity
type Configuration struct {
	ID          string              `json:"id"`
	Name        string              `json:"name"`
	Type        ConfigurationType   `json:"type"`
	ParentID    *string             `json:"parent_id"`
	Data        json.RawMessage     `json:"data"`
	CreatedBy   string              `json:"created_by"`
	Description *string             `json:"description"`
	Status      ConfigurationStatus `json:"status"`
	Archived    bool                `json:"archived"`
	CreatedAt   time.Time           `json:"created_at"`
	UpdatedAt   time.Time           `json:"updated_at"`
}

// GetConfigurationsOptions represents options for getting configurations
type GetConfigurationsOptions struct {
	Type            *ConfigurationType   `json:"type,omitempty"`
	Status          *ConfigurationStatus `json:"status,omitempty"`
	IncludeArchived *bool                `json:"includeArchived,omitempty"`
}

// ConfigurationListResponse represents configuration list response
type ConfigurationListResponse struct {
	Configs []Configuration `json:"configs"`
}

// CreateConfigurationRequest represents create configuration request
type CreateConfigurationRequest struct {
	Name        string            `json:"name"`
	Type        ConfigurationType `json:"type"`
	ParentID    *string           `json:"parent_id,omitempty"`
	Data        interface{}       `json:"data"`
	Description *string           `json:"description,omitempty"`
}

// UpdateConfigurationRequest represents update configuration request
type UpdateConfigurationRequest struct {
	Data        interface{} `json:"data,omitempty"`
	Description *string     `json:"description,omitempty"`
}

// ConfigurationResponse represents configuration response with message
type ConfigurationResponse struct {
	Config  Configuration `json:"config"`
	Message *string       `json:"message,omitempty"`
}

// ResolvedConfigurationResponse represents resolved configuration response
type ResolvedConfigurationResponse struct {
	Config           Configuration          `json:"config"`
	Data             json.RawMessage        `json:"data"`
	InheritanceChain []ConfigurationSummary `json:"inheritance_chain,omitempty"`
}

// ConfigurationSummary represents configuration summary for inheritance chain
type ConfigurationSummary struct {
	ID   string            `json:"id"`
	Name string            `json:"name"`
	Type ConfigurationType `json:"type"`
}

// GetConfigurationOptions represents options for getting a configuration
type GetConfigurationOptions struct {
	Provenance *bool `json:"provenance,omitempty"`
	Raw        *bool `json:"raw,omitempty"`
}

// ConfigurationValueResponse represents configuration value response
type ConfigurationValueResponse struct {
	Value  json.RawMessage        `json:"value"`
	Path   *string                `json:"path,omitempty"`
	Source *ConfigurationSource   `json:"source,omitempty"`
	Config *ConfigurationSummary  `json:"config,omitempty"`
	Data   *json.RawMessage       `json:"data,omitempty"`
}

// GetConfigurationValueOptions represents options for getting configuration value
type GetConfigurationValueOptions struct {
	Path    *string `json:"path,omitempty"`
	Minimal *bool   `json:"minimal,omitempty"`
}

// ConfigurationSource represents configuration source information
type ConfigurationSource struct {
	ID        string             `json:"id"`
	Name      string             `json:"name"`
	Type      ConfigurationType  `json:"type"`
	CreatedBy *string            `json:"createdBy,omitempty"`
	CreatedAt *time.Time         `json:"createdAt,omitempty"`
}

// ChildrenConfigurationResponse represents children configurations response
type ChildrenConfigurationResponse struct {
	Children []Configuration `json:"children"`
	Count    int             `json:"count"`
}

// GetChildConfigurationsOptions represents options for getting child configurations
type GetChildConfigurationsOptions struct {
	IncludeArchived *bool `json:"includeArchived,omitempty"`
}

// ComponentsResponse represents components with versions response
type ComponentsResponse struct {
	Components []ComponentWithVersions `json:"components"`
}

// ComponentWithVersions represents a component with its versions
type ComponentWithVersions struct {
	ID          string              `json:"id"`
	Name        string              `json:"name"`
	Type        ConfigurationType   `json:"type"`
	Description *string             `json:"description"`
	Status      ConfigurationStatus `json:"status"`
	CreatedBy   string              `json:"created_by"`
	CreatedAt   time.Time           `json:"created_at"`
	Versions    []Configuration     `json:"versions"`
}

// RenameConfigurationRequest represents rename configuration request
type RenameConfigurationRequest struct {
	Name string `json:"name"`
}

// ArchiveConfigurationRequest represents archive configuration request
type ArchiveConfigurationRequest struct {
	ArchiveChildren *bool `json:"archiveChildren,omitempty"`
}

// ConfigurationService provides configuration management operations
type ConfigurationService struct {
	client *HTTPClient
}

// NewConfigurationService creates a new configuration service
func NewConfigurationService(client *HTTPClient) *ConfigurationService {
	return &ConfigurationService{client: client}
}

File: configuration-manager-python-client/configuration_manager_client/models/configuration.py
"""Configuration models for Configuration Manager client."""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Any, Dict, List

from .types import ConfigurationType, ConfigurationStatus


@dataclass
class Configuration:
    """Configuration entity."""
    id: str
    name: str
    type: ConfigurationType
    parent_id: Optional[str]
    data: Dict[str, Any]
    created_by: str
    description: Optional[str]
    status: ConfigurationStatus
    archived: bool
    created_at: datetime
    updated_at: datetime


@dataclass
class GetConfigurationsOptions:
    """Options for getting configurations."""
    type: Optional[ConfigurationType] = None
    status: Optional[ConfigurationStatus] = None
    include_archived: Optional[bool] = None


@dataclass
class ConfigurationListResponse:
    """Configuration list response."""
    configs: List[Configuration]


@dataclass
class CreateConfigurationRequest:
    """Create configuration request."""
    name: str
    type: ConfigurationType
    data: Dict[str, Any]
    parent_id: Optional[str] = None
    description: Optional[str] = None


@dataclass
class UpdateConfigurationRequest:
    """Update configuration request."""
    data: Optional[Dict[str, Any]] = None
    description: Optional[str] = None


@dataclass
class ConfigurationResponse:
    """Configuration response with message."""
    config: Configuration
    message: Optional[str] = None


@dataclass
class ConfigurationSummary:
    """Configuration summary for inheritance chain."""
    id: str
    name: str
    type: ConfigurationType


@dataclass
class ResolvedConfigurationResponse:
    """Resolved configuration response."""
    config: Configuration
    data: Dict[str, Any]
    inheritance_chain: Optional[List[ConfigurationSummary]] = None


@dataclass
class GetConfigurationOptions:
    """Options for getting a configuration."""
    provenance: Optional[bool] = None
    raw: Optional[bool] = None


@dataclass
class ConfigurationSource:
    """Configuration source information."""
    id: str
    name: str
    type: ConfigurationType
    created_by: Optional[str] = None
    created_at: Optional[datetime] = None


@dataclass
class ConfigurationValueResponse:
    """Configuration value response."""
    value: Any
    path: Optional[str] = None
    source: Optional[ConfigurationSource] = None
    config: Optional[ConfigurationSummary] = None
    data: Optional[Dict[str, Any]] = None


@dataclass
class GetConfigurationValueOptions:
    """Options for getting configuration value."""
    path: Optional[str] = None
    minimal: Optional[bool] = None


@dataclass
class ChildrenConfigurationResponse:
    """Children configurations response."""
    children: List[Configuration]
    count: int


@dataclass
class GetChildConfigurationsOptions:
    """Options for getting child configurations."""
    include_archived: Optional[bool] = None


@dataclass
class ComponentWithVersions:
    """Component with its versions."""
    id: str
    name: str
    type: ConfigurationType
    description: Optional[str]
    status: ConfigurationStatus
    created_by: str
    created_at: datetime
    versions: List[Configuration] = field(default_factory=list)


@dataclass
class ComponentsResponse:
    """Components with versions response."""
    components: List[ComponentWithVersions]


@dataclass
class RenameConfigurationRequest:
    """Rename configuration request."""
    name: str


@dataclass
class ArchiveConfigurationRequest:
    """Archive configuration request."""
    archive_children: Optional[bool] = None
